extra_data/reader.py

```diff
@@ -104,14 +105,15 @@ def __init__(self, filename, _cache_info=None):
 
     @property
     def file(self):
-        if self._file is None:
-            self._file = h5py.File(self.filename, 'r')
+        if self._file.id.valid:
+            get_global_filecache().touch(self.filename)
+        else:
+            self._file = get_global_filecache().get_or_open(self.filename)
+            
         return self._file
 
     def close(self):
-        if self._file:
-            self._file.close()
-            self._file = None
+        pass
```

Comment:
I think it should still be possible to close files explicitly, removing them from the cache as we do so. If I do:

```
with RunDirectory('...') as run:
    run.info()
```

I expect that to close any files deterministically when the block finishes.

There's some ugliness here because the same files might be open in another DataCollection, so closing them might be suboptimal. But if we have to choose one or the other, I'd prefer to be a bit too eager closing files rather than leaving open files lying around.

It may be possible to avoid having to choose one or the other, by keeping weakrefs in the cache so they automatically expire when no other references to them exist.

Reply:
Even with `weakref` we cannot guarantee that the underlying file will be actually close. So it doesn't make sense for me to be eager here.
If it is nessecary to actually close the file I implemented methods `clean` and `close` in `FileCache`

Reply:
Looking at the code in h5py, I think files which drop out of scope will be closed correctly (it uses [`H5Idec_ref`](https://confluence.hdfgroup.org/display/HDF5/H5I_DEC_REF) on them). 

But as I said, if we're not confident we can do that properly, I would prefer to close files and potentially make some code re-open them rather than leaving unwanted file handles open.

Reply:
In fact, h5py never calls `H5Fclose`. `file.close()` finds any objects belonging to the file, calls `H5Idec_ref` on them, and then calls `H5Idec_ref` on the file ID itself - the same function that's called when a file goes out of scope. So I think we could rely on reference counting for the cleanup. I'll open an issue on h5py to clarify this.

Reply:
I actually close explicitly all files in cache at the exit:
```
@atexit.register
def clean_global_filecache():
    _extra_data_file_cache.clean()
```

and I will add the __del__ method in FileCache then:
```
def __del__(self):
    self.clean()
```

I hope we can reset the cache without closing all files after the process fork in multiprocessing mode.

Reply:
right, but `atexit` is only when the entire process exits, at which point our read-only file handles are going to get closed anyway. I still maintain that it should be possible to open a run using something like a `with` block, and have the files actually closed when you'd expect, even with the cache.

Reply:
`weakref` will require `WeakValueOrderedDict` implementation.  As an alternative to this bad idea, I can add a counter of `FileAccess` instances referred to a particular cached file. It will require to keep an `if` in `FileAccess.file`, and an additional method for soft close in `FileCache`. Not so complex as a full  `WeakValueOrderedDict`. But it doesn't make sense. 

We cannot guarantee that files after `with` block will be actually closed using a soft close. It means we must write in the documentation:
> Never expect that underling h5 files closed after destroying `DataCollection` instance. If you need forcefully close a file use `get_global_filecache().close(filename)`.

And it means that users must use this call anyway if they need to close a file or they may not care at all. Cleaning at `FileCache.__del__()` guarantee that all files will be closed properly at the end.   

So we are going waste efforts and make code more complex.

Reply:
I'll keep thinking about the best way to do it, but I do think the idea is important: whatever caching we add shouldn't keep files open after they would naturally be closed. I hope we can find a way to do that without a massive amount of extra complexity, but if it means some extra complexity, I think it's worth doing.

Reply:
Do you think that you are following the concept of isolation, which in fact should be accepted as a dogma in most cases. But in our case, it has no effect from the moment we make the file cache global.

However, to close our discussion, I implemented soft closing of files as you want. Although this is unnecessary in my opinion, in principle, it does not violate the invariant.